### PR TITLE
Implement profiles, container configuration, and devices

### PR DESCRIPTION
**Pull request by hallyn**
For devices, only veth network interfaces and bind mounts are implemented.

Signed-off-by: Serge Hallyn <serge.hallyn@ubuntu.com>

### CODE REVIEW

File: lxc/config.go [hunk 1]
```diff
@@ -56,7 +101,12 @@ func (c *configCmd) run(config *lxd.Config, args []string) error {
 			return err
 		}
 
-		return fmt.Errorf(gettext.Gettext("Only 'password' can be set currently"))
+		if len(args) < 3 {
+			return fmt.Errorf(gettext.Gettext("Bad config command"))
```
Comment (tych0) [inline lxc/config.go]: Can we return `errArgs` in all these cases? That way lxc will print out the help() for the command when someone does something wrong.

File: lxc/config.go [hunk 2]
```diff
@@ -141,7 +191,341 @@ func (c *configCmd) run(config *lxd.Config, args []string) error {
...
+		remote = config.DefaultRemote
+	}
+
+	client, err := lxd.NewClient(config, remote)
+	if err != nil {
+		return err
+	}
+
+	/*
+	var filters []string
```
Comment (tych0) [inline lxc/config.go]: Can we get rid of this code?

File: lxc/config.go [hunk 3]
```diff
@@ -141,7 +191,341 @@ func (c *configCmd) run(config *lxd.Config, args []string) error {
...
+	}
+
+	var resp *lxd.Response
+	if which == "profile" {
+		resp, err = client.ProfileDeviceAdd(name, devname, devtype, props)
+	} else {
+		resp, err = client.ContainerDeviceAdd(name, devname, devtype, props)
+	}
+	if err != nil {
+		fmt.Printf(gettext.Gettext("Error: %s\n"), err)
```
Comment (tych0) [inline lxc/config.go]: I think we can just return the error here and it'll get printed out for us.

File: lxd/containers.go [hunk 4]
```diff
@@ -369,6 +369,165 @@ func containerDeleteSnapshots(d *Daemon, cname string) []string {
...
+	_, err = tx.Exec(`DELETE FROM containers_devices_config WHERE id IN
+		(SELECT containers_devices_config.id
+		 FROM containers_devices_config JOIN containers_devices
+		 ON containers_devices_config.container_device_id=containers_devices.id
+		 WHERE containers_devices.container_id=?)`, id)
+	if err != nil {
+		return err
+	}
+	_, err = tx.Exec("DELETE FROM containers_devices WHERE container_id=?", id)
+	if err != nil {
```
Comment (tych0) [inline lxd/containers.go]: I think you can just `return err` here.

File: lxd/containers.go [hunk 5]
```diff
@@ -369,6 +369,165 @@ func containerDeleteSnapshots(d *Daemon, cname string) []string {
...
+func ValidContainerConfigKey(k string) bool {
+	switch k {
+	case "limits.cpus": return true
+	case "limits.memory": return true
+	case "security.privileged": return true
+	case "raw.apparmor": return true
+	case "raw.lxc": return true
+	}
+
+	if strings.HasPrefix(k, "user.") {
```
Comment (tych0) [inline lxd/containers.go]: Can you just `return strings.HasPrefix(k, "user.")` here?

File: lxd/containers.go [hunk 6]
```diff
@@ -369,6 +369,165 @@ func containerDeleteSnapshots(d *Daemon, cname string) []string {
...
+			}
+		}
+
+		err = shared.AddDevices(tx, "container", cId, configRaw.Devices)
+		if err != nil {
+			tx.Rollback()
+			return err
+		}
+
+		if err := tx.Commit(); err != nil {
```
Comment (tych0) [inline lxd/containers.go]: You can `return tx.Commit()` here.

File: lxd/containers.go [hunk 7]
```diff
@@ -440,6 +602,14 @@ func (c *lxdContainer) Freeze() error {
 	return c.c.Freeze()
 }
 
+func (c *lxdContainer) isUnprivileged() bool {
+	switch c.config["security.privileged"] {
+	case "0": return true
+	case "false": return true
```
Comment (tych0) [inline lxd/containers.go]: Should we lowercase the input here to match False and false?

File: lxd/db.go [hunk 8]
```diff
@@ -253,3 +375,126 @@ func dbGetConfig(d *Daemon, c *lxdContainer) (map[string]string, error) {
...
+		var order int
+		var p Profile
+		if err := rows.Scan(&name, &order); err != nil {
+			return nil, err
+		}
+		p.name = name
+		p.order = order
+		profiles = append(profiles, p)
+	}
+	sort.Sort(profiles)
```
Comment (tych0) [inline lxd/db.go]: Should we use ORDER BY here or sort it on the go side once we're done?

File: client.go [hunk 9]
```diff
@@ -1019,3 +1042,329 @@ func (c *Client) ListSnapshots(container string) ([]string, error) {
...
+ * return string array representing a container's full configuration
+ */
+func (c *Client) GetContainerConfig(container string) ([]string, error) {
+	st, err := c.ContainerStatus(container)
+	var resp []string
+	if err != nil {
+		return resp, err
+	}
+
+	profiles := strings.Join(st.Profiles, ",")
```
Comment (tych0) [inline client.go]: Can we make this `", "` since its for human consumption?

File: lxc/config.go (hunk 2 above)
Comment (hallyn) [inline lxc/config.go]: Do you mean the commented out filters code?

Comment (tych0) [inline lxc/config.go]: On Thu, Mar 05, 2015 at 10:07:22AM -0800, Serge Hallyn wrote:
> > +		var name string
> > +		remote, name = config.ParseRemoteAndContainer(args[2])
> > +		if name != "" {
> > +			return fmt.Errorf("Cannot provide container name to list")
> > +		}
> > +	} else {
> > +		remote = config.DefaultRemote
> > +	}
> > +
> > +	client, err := lxd.NewClient(config, remote)
> > +	if err != nil {
> > +		return err
> > +	}
> > +
> > +	/*
> > +	var filters []string
> 
> Do you mean the commented out filters code?

Yep.

File: client.go (hunk 9 above)
Comment (hallyn) [inline client.go]: I don't think so, bc ppl may then cut-paste into a profile apply, and spaces are not ok there.

Comment (tych0) [inline client.go]: On Thu, Mar 05, 2015 at 10:21:39AM -0800, Serge Hallyn wrote:
> > @@ -1019,3 +1042,329 @@ func (c *Client) ListSnapshots(container string) ([]string, error) {
> >  
> >  	return names, nil
> >  }
> > +
> > +/*
> > + * return string array representing a container's full configuration
> > + */
> > +func (c *Client) GetContainerConfig(container string) ([]string, error) {
> > +	st, err := c.ContainerStatus(container)
> > +	var resp []string
> > +	if err != nil {
> > +		return resp, err
> > +	}
> > +
> > +	profiles := strings.Join(st.Profiles, ",")
> 
> I don't think so, bc ppl may then cut-paste into a profile apply, and spaces are not ok there.

Ah, ok.

**Pull request merged**